test/geometry/test_bbox_v2.py

```diff
@@ -0,0 +1,333 @@
+
+    @pytest.mark.parametrize('shape', [(1, 4), (1, 1, 4)])
+    def test_bounding_boxes_convert_to_kornia(self, shape: Tuple[int], device, dtype):
+        box_xyxy = torch.as_tensor([[1, 2, 3, 4]], device=device, dtype=dtype).view(*shape)
+        box_xyxy_plus_1 = torch.as_tensor([[1, 2, 4, 5]], device=device, dtype=dtype).view(*shape)
+        box_xywh = torch.as_tensor([[1, 2, 2, 2]], device=device, dtype=dtype).view(*shape)
+        box_xywh_plus_1 = torch.as_tensor([[1, 2, 3, 3]], device=device, dtype=dtype).view(*shape)
+
+        expected_box = torch.as_tensor([[[1, 2], [3, 2], [3, 4], [1, 4]]], device=device, dtype=dtype).view(*shape, 2)
+
+        kornia_xyxy = bbox_to_kornia_bbox(box_xyxy, mode='xyxy')
+        kornia_xyxy_plus_1 = bbox_to_kornia_bbox(box_xyxy_plus_1, mode='xyxy_plus_1')
+        kornia_xywh = bbox_to_kornia_bbox(box_xywh, mode='xywh')
+        kornia_xywh_plus_1 = bbox_to_kornia_bbox(box_xywh_plus_1, mode='xywh_plus_1')
+
+        assert kornia_xyxy.shape == expected_box.shape
+        assert_allclose(kornia_xyxy, expected_box)
+
+        assert kornia_xyxy_plus_1.shape == expected_box.shape
+        assert_allclose(kornia_xyxy_plus_1, expected_box)
+
+        assert kornia_xywh.shape == expected_box.shape
+        assert_allclose(kornia_xywh, expected_box)
+
+        assert kornia_xywh_plus_1.shape == expected_box.shape
+        assert_allclose(kornia_xywh_plus_1, expected_box)
+
+    @pytest.mark.parametrize('shape', [(1, 4), (1, 1, 4)])
+    def test_bounding_boxes_convert_from_kornia(self, shape: Tuple[int], device, dtype):
+        box = torch.as_tensor([[[1, 2], [3, 2], [3, 4], [1, 4]]], device=device, dtype=dtype).view(*shape, 2)
+
+        expected_box_xyxy = torch.as_tensor([[1, 2, 3, 4]], device=device, dtype=dtype).view(*shape)
+        expected_box_xyxy_plus_1 = torch.as_tensor([[1, 2, 4, 5]], device=device, dtype=dtype).view(*shape)
+        expected_box_xywh = torch.as_tensor([[1, 2, 2, 2]], device=device, dtype=dtype).view(*shape)
+        expected_box_xywh_plus_1 = torch.as_tensor([[1, 2, 3, 3]], device=device, dtype=dtype).view(*shape)
+
+        kornia_xyxy = kornia_bbox_to_bbox(box, mode='xyxy')
+        kornia_xyxy_plus_1 = kornia_bbox_to_bbox(box, mode='xyxy_plus_1')
+        kornia_xywh = kornia_bbox_to_bbox(box, mode='xywh')
+        kornia_xywh_plus_1 = kornia_bbox_to_bbox(box, mode='xywh_plus_1')
+
+        assert kornia_xyxy.shape == expected_box_xyxy.shape
+        assert_allclose(kornia_xyxy, expected_box_xyxy)
+
+        assert kornia_xyxy_plus_1.shape == expected_box_xyxy_plus_1.shape
+        assert_allclose(kornia_xyxy_plus_1, expected_box_xyxy_plus_1)
+
+        assert kornia_xywh.shape == expected_box_xywh.shape
+        assert_allclose(kornia_xywh, expected_box_xywh)
+
+        assert kornia_xywh_plus_1.shape == expected_box_xywh_plus_1.shape
+        assert_allclose(kornia_xywh_plus_1, expected_box_xywh_plus_1)
+
+    def test_gradcheck(self, device, dtype):
```

Comment:
Fails. I don't know how to solve it. It failed for the old kornia.geometry.bbox code.

```
test/geometry/test_bbox_v2.py:122 (TestBbox2D.test_gradcheck[cpu-float32])
self = <test.geometry.test_bbox_v2.TestBbox2D object at 0x7f128c56ff90>
device = device(type='cpu'), dtype = torch.float32

    def test_gradcheck(self, device, dtype):
        boxes1 = torch.tensor([[[1.0, 1.0], [3.0, 1.0], [3.0, 2.0], [1.0, 2.0]]], device=device, dtype=dtype)
        boxes1 = utils.tensor_to_gradcheck_var(boxes1)
        boxes2 = utils.tensor_to_gradcheck_var(boxes1.detach().clone())
    
        boxes_xyxy = torch.tensor([[1.0, 3.0, 5.0, 6.0]])
    
>       assert gradcheck(infer_bbox_shape, (boxes1,), raise_exception=True)

test_bbox_v2.py:130: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.dev_env/envs/venv/lib/python3.7/site-packages/torch/autograd/gradcheck.py:1245: in gradcheck
    return _gradcheck_helper(**args)
../../.dev_env/envs/venv/lib/python3.7/site-packages/torch/autograd/gradcheck.py:1259: in _gradcheck_helper
    rtol, atol, check_grad_dtypes, check_forward_ad=check_forward_ad, nondet_tol=nondet_tol)
../../.dev_env/envs/venv/lib/python3.7/site-packages/torch/autograd/gradcheck.py:931: in _gradcheck_real_imag
    rtol, atol, check_grad_dtypes, nondet_tol)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

func = <function infer_bbox_shape at 0x7f128cde4f80>
func_out = (tensor([1.], dtype=torch.float64, grad_fn=<SelectBackward>), tensor([2.], dtype=torch.float64, grad_fn=<SelectBackward>))
tupled_inputs = (tensor([[[1., 1.],
         [3., 1.],
         [3., 2.],
         [1., 2.]]], dtype=torch.float64, grad_fn=<CopyBackwards>),)
outputs = (tensor([1.], dtype=torch.float64, grad_fn=<SelectBackward>), tensor([2.], dtype=torch.float64, grad_fn=<SelectBackward>))
eps = 1e-06, rtol = 0.001, atol = 1e-05, check_grad_dtypes = False
nondet_tol = 0.0

    def _slow_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol, atol, check_grad_dtypes,
                        nondet_tol, *, use_forward_ad=False, complex_indices=None, test_imag=False):
        if not outputs:
            return _check_no_differentiable_outputs(func, tupled_inputs, _as_tuple(func_out), eps)
    
        numerical = _transpose(_get_numerical_jacobian(func, tupled_inputs, outputs, eps=eps, is_forward_ad=use_forward_ad))
    
        if use_forward_ad:
            analytical_forward = _get_analytical_jacobian_forward_ad(func, tupled_inputs, outputs, check_grad_dtypes=check_grad_dtypes)
    
            for i, n_per_out in enumerate(numerical):
                for j, n in enumerate(n_per_out):
                    a = analytical_forward[j][i]
                    if not _allclose_with_type_promotion(a, n.to(a.device), rtol, atol):
                        raise GradcheckError(_get_notallclose_msg(a, n, i, j, complex_indices, test_imag,
                                                                  is_forward_ad=True))
        else:
            for i, o in enumerate(outputs):
                analytical = _check_analytical_jacobian_attributes(tupled_inputs, o, nondet_tol, check_grad_dtypes)
    
                for j, (a, n) in enumerate(zip(analytical, numerical[i])):
                    if not _allclose_with_type_promotion(a, n.to(a.device), rtol, atol):
>                       raise GradcheckError(_get_notallclose_msg(a, n, i, j, complex_indices, test_imag))
E                       torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
E                       numerical:tensor([[ 0.0000],
E                               [-0.5000],
E                               [ 0.0000],
E                               [-0.5000],
E                               [ 0.0000],
E                               [ 0.5000],
E                               [ 0.0000],
E                               [ 0.5000]], dtype=torch.float64)
E                       analytical:tensor([[ 0.],
E                               [-1.],
E                               [ 0.],
E                               [ 0.],
E                               [ 0.],
E                               [ 1.],
E                               [ 0.],
E                               [ 0.]], dtype=torch.float64)

../../.dev_env/envs/venv/lib/python3.7/site-packages/torch/autograd/gradcheck.py:978: GradcheckError

```

Reply:
@shijianjian @dkoguciuk any idea about this ?

Reply:
I think functions like ```infer_bbox_shape``` should be non-differentiable. The gradients of shape inferring does not make too much sense.